wirecard/oxid-ee
# WDPD-168 Backend Metadata Table

### This PR

* Adds meta data to payments
* Creates a new table to store meta data values
* Adds currency multiselect field for Ratepay Invoice payment method
* Adds a database helper

### Notes

The meta data extension is necessary so that arbitrary data can be stored to any model. In OXID, you can only store data that has a representation in the database, i.e. a column exists in the table associated with a model. All data from these columns is made available via magic getters and setters.

For example: The Payment model's associated table is `oxpayments`. It has multiple columns, e.g. `OXACTIVE` or `OXDESC`. Data can be retrieved from the database by calling `$oPayment->oxpayments__oxactive`, `$oPayment->oxpayments__oxdesc` and so on. It can be saved by setting these properties and then calling `save()` on the object.

With this extension, this functionality is no longer tied to columns in the database. So, for example, you can also set `$oPayment->oxpayments__something` and it will be stored in the database.

### How to Test

* Activate the module
* Go to _Shop Settings → Payment Methods_ and activate _Wirecard Garantierter Kauf auf Rechnung_
* Confirm that there is a setting for "Currency" and that multiple values can be selected
* Click "Save" and check if the selected currencies are saved
* For the curious: Connect to the database, select the table `wdoxidee_oxpayments_metadata` and confirm that there is an entry for the selected currencies

### Jira Links

* WDPD-164: https://jira.parkside.at/browse/WDPD-164
* WDPD-168: https://jira.parkside.at/browse/WDPD-168
